### PR TITLE
Fix: Do not emit health_status event for each health check attempt.

### LINKED ISSUE #24003: health_status events are too noisy/redundant
### Issue Description

1. the `health_status` events emitted by podman is too noisy.
    - That is, `health_status` event is emitted for each health check attempts . I think it should only be emitted when ever there is a change in `health_status` status of a container. 
2. Docker handles this case in better way. Docker is emitting different events (`exec_create` and `exec_die`) for  health check attempts and only emit `health_status` event whenever there is an actual status/state change.
    - In the above way other applications which implements docker/podman service discovery  can safely and easly depend on docker/podman for service discovery 
    - For eg: tools like  Traefik proxy implements docker/podman service discovery by reloading confiuration for every health_status event. 
    - Because of the noisy behaviour,  there will be considerable CPU resource utilization if we try to use tools like Traefik along with podman. 
    - I am not raising this bug report just because podman is not implementing  docker's behaviour, but I believe docker is handling this case in better way than podman

As a secure  container runtime, I see very big opportunity  for Podman . Bugs like this will prevent people from switching to podman . 
I found this issue while using Coolify along with podman where considerable CPU resource is wasted due to this bug 

### Steps to reproduce the issue

Steps to reproduce the issue
1. clone https://github.com/harish2704/podman-health-check-bug. 
2. run `cd for-podman`
3. run `podman-compose up -d`
4. watch events emitted by podman by running `podman events --format json | jq -r '[.time, .Name, .Status, .health_status] | @tsv'`

### Describe the results you received

The output I am seeing is given below.
```
1726677092      for-podman_site1_1      health_status   healthy
1726677092      for-podman_site0_1      health_status   healthy
1726677093      for-podman_site2_1      health_status   healthy
1726677093      for-podman_site1_1      health_status   healthy
1726677093      for-podman_site0_1      health_status   healthy
1726677094      for-podman_site2_1      health_status   healthy
1726677094      for-podman_site1_1      health_status   healthy
1726677094      for-podman_site0_1      health_status   healthy
1726677096      for-podman_site1_1      health_status   healthy
1726677096      for-podman_site0_1      health_status   healthy
1726677096      for-podman_site2_1      health_status   healthy
1726677098      for-podman_site2_1      health_status   healthy
1726677098      for-podman_site0_1      health_status   healthy
1726677098      for-podman_site1_1      health_status   healthy
1726677100      for-podman_site2_1      health_status   healthy
1726677102      for-podman_site0_1      health_status   healthy
1726677102      for-podman_site1_1      health_status   healthy
1726677103      for-podman_site2_1      health_status   healthy
1726677103      for-podman_site0_1      health_status   healthy
1726677103      for-podman_site1_1      health_status   healthy
1726677105      for-podman_site2_1      health_status   unhealthy
1726677105      for-podman_site1_1      health_status   unhealthy
1726677105      for-podman_site0_1      health_status   unhealthy
1726677106      for-podman_site1_1      health_status   unhealthy
1726677106      for-podman_site2_1      health_status   unhealthy
1726677107      for-podman_site0_1      health_status   unhealthy
1726677108      for-podman_site0_1      health_status   unhealthy
1726677108      for-podman_site1_1      health_status   unhealthy
1726677108      for-podman_site2_1      health_status   unhealthy
1726677110      for-podman_site2_1      health_status   unhealthy
```

In the above , we can see that most of the  `health_status` events are duplicates. ( doesn't contain any new information )

### Describe the results you expected

I am expecting the `health_status` event only when there is a status/change change 

### podman info output

```yaml
host:
  arch: amd64
  buildahVersion: 1.37.0
  cgroupControllers:
  - cpu
  - io
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.12-2.fc40.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.12, commit: '
  cpuUtilization:
    idlePercent: 96.27
    systemPercent: 1.64
    userPercent: 2.09
  cpus: 12
  databaseBackend: sqlite
  distribution:
    distribution: fedora
    variant: kde
    version: "40"
  eventLogger: journald
  freeLocks: 2012
  hostname: fedora-desk
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 524288
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 524288
      size: 65536
  kernel: 6.10.9-200.fc40.x86_64
  linkmode: dynamic
  logDriver: journald
  memFree: 337436672
  memTotal: 16680095744
  networkBackend: netavark
  networkBackendInfo:
    backend: netavark
    dns:
      package: aardvark-dns-1.12.2-2.fc40.x86_64
      path: /usr/libexec/podman/aardvark-dns
      version: aardvark-dns 1.12.2
    package: netavark-1.12.2-1.fc40.x86_64
    path: /usr/libexec/podman/netavark
    version: netavark 1.12.2
  ociRuntime:
    name: crun
    package: crun-1.17-1.fc40.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.17
      commit: 000fa0d4eeed8938301f3bcf8206405315bc1017
      rundir: /run/user/1000/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +LIBKRUN +WASM:wasmedge +YAJL
  os: linux
  pasta:
    executable: /usr/bin/pasta
    package: passt-0^20240906.g6b38f07-1.fc40.x86_64
    version: |
      pasta 0^20240906.g6b38f07-1.fc40.x86_64
      Copyright Red Hat
      GNU General Public License, version 2 or later
        <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
      This is free software: you are free to change and redistribute it.
      There is NO WARRANTY, to the extent permitted by law.
  remoteSocket:
    exists: true
    path: /run/user/1000/podman/podman.sock
  rootlessNetworkCmd: pasta
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: ""
    selinuxEnabled: true
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.2.2-2.fc40.x86_64
    version: |-
      slirp4netns version 1.2.2
      commit: 0ee2d87523e906518d34a6b423271e4826f71faf
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.5
  swapFree: 7665238016
  swapTotal: 8589930496
  uptime: 28h 59m 43.00s (Approximately 1.17 days)
  variant: ""
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries: {}
store:
  configFile: /home/harish/.config/containers/storage.conf
  containerStore:
    number: 20
    paused: 0
    running: 4
    stopped: 16
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/harish/.local/share/containers/storage
  graphRootAllocated: 236221104128
  graphRootUsed: 91409645568
  graphStatus:
    Backing Filesystem: btrfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Supports shifting: "false"
    Supports volatile: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 50
  runRoot: /run/user/1000/containers
  transientStore: false
  volumePath: /home/harish/.local/share/containers/storage/volumes
version:
  APIVersion: 5.3.0-dev
  Built: 1726676363
  BuiltTime: Wed Sep 18 21:49:23 2024
  GitCommit: 62c101651ff85daa0370d5031b9e2b3b4c5f16be
  GoVersion: go1.22.6
  Os: linux
  OsArch: linux/amd64
  Version: 5.3.0-dev
```
```


### Podman in a container

No

### Privileged Or Rootless

Rootless

### Upstream Latest Release

Yes

### Additional environment details

My OS
```
NAME="Fedora Linux"
VERSION="40 (KDE Plasma)"
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by harish2704**
Emit `health_status` event only if there is a change in health_status
Fixes #24003

#### Does this PR introduce a user-facing change?

```release-note
Emit `health_status` event only when there a status/state change for `healthy` flag
```

### CODE REVIEW

**Comment by martinpitt** [comment]
This "breaks" cockpit-podman's testHealthcheckSystem, which assumes the current behavior of getting regular "health check ran and passed" events. I don't have a good gut feeling whether the regular "pings of life" are expected/by design or considered noise. So I'll defer to the review of the podman developers, and if this change is approved, I'll adjust cockpit-podman's tests.

File: libpod/healthcheck.go [hunk 1]
```diff
@@ -397,15 +398,19 @@ func (c *Container) updateHealthCheckLog(hcl define.HealthCheckLog, inStartPerio
...
 	if len(healthCheck.Log) > MaxHealthCheckNumberLogs {
 		healthCheck.Log = healthCheck.Log[1:]
 	}
 	newResults, err := json.Marshal(healthCheck)
 	if err != nil {
-		return "", fmt.Errorf("unable to marshall healthchecks for writing: %w", err)
+		return "", isHCStausChanged, fmt.Errorf("unable to marshall healthchecks for writing: %w", err)
+	}
+	if isHCStausChanged {
+		err = os.WriteFile(c.healthCheckLogPath(), newResults, 0700)
```
Comment (Luap99) [inline libpod/healthcheck.go:411]: That is not correct, the log contains timestamp and output of the command and that must be logged for each healthcheck as it is displayed in podman inspect.

**Review by Luap99 — CHANGES_REQUESTED** [review]
> I don't have a good gut feeling whether the regular "pings of life" are expected/by design or considered noise

I don't think there is a formal design on how it should work but the way it works today is how the current users will expect it to work. The fact that it breaks cockpit testing is a good sign that we have users depending on it. As such I consider this a breaking change that is not suitable for a minor version so this would have to wait for podman 6.0 if we want to do that at all. I think reducing the event spam is a good idea in general.

Now one thing we should consider if docker doesn't behave this way our docker compat api should not behave this way either. One way would be to a new field to the health_status event that is set when the status changed and then we can filter out the events that did not have this set to make the docker clients work correctly at least.

cc @mheon @Honny1 In case you have opinions as you have been working on other healthcheck events work

File: libpod/healthcheck.go (hunk 1 above)
Comment (harish2704) [reply libpod/healthcheck.go:411]: Got it. I thought  `healthCheckLogPath` is a persistent log of all `health_status` events. 
I will correct it once we finalize how we can move forward.

**Review by Honny1 — COMMENTED** [review]
Why not, reduce the noise in the logs, but I would wait for Podman 6.0. I'd also be in favor of adding a flag to enable log "saving" for each run, for `podman healtcheck run` (probably also for `podman run`), for debugging purposes or if something goes wrong.

**Comment by harish2704** [comment]
@Luap99 
> I don't think there is a formal design on how it should work but the way it works today is how the current users will expect it to work.

I have to agree on this because I searched for such a documentation ( https://docs.docker.com/reference/api/engine/version/v1.41/#tag/System/operation/SystemEvents ) and there is no description about exact behavior of those events . They have only provided the list of valid events 

> The fact that it breaks cockpit testing is a good sign that we have users depending on it. As such I consider this a breaking change that is not suitable for a minor version so this would have to wait for podman 6.0 if we want to do that at all

IMHO,  Podman has a huge opportunity as secure docker replacement and in that sense most of the users ( and applications targeting docker eg: Traefik ) will expect it work similar to docker. To address the user base who is looking to switch to Podman, it should be considered as a bug. Please note that, here i am not referring the `podman` cli tool. I am talking about docker compatible API which podman provides. 

> Now one thing we should consider if docker doesn't behave this way our docker compat api should not behave this way either. One way would be to a new field to the health_status event that is set when the status changed and then we can filter out the events that did not have this set to make the docker clients work correctly at least.

Exactly. This is the point I was trying to convey. I checked how `podman-docker` works but it is found to be a simple bash wrapper script for podman cli ( at-least in my Fedora-40 ) . 
So, If my above understanding is correct ( ie, podman-docker is just an alias/wrapper of podman ) then, `podman-docker` is not exposing a separate  docker compatible API ( either via unix socket or via TCP ) . 
All we have is an API exposed by podman and it is expected to be docker compatible. 

So, in short, my bug report is  not about behavior of `podma-cli` tool. it is about the docker-compatible API which podman exposes.  
if any of you can provide some hints on how to fix this issue without changing behaviour `podman` cli I am happy to try that. To be specific, my question is , at which point, Golang event is converted to HTTP API data? 
I will try to add the additional flag as you mentioned in the comment

**Comment by mheon** [comment]
IMO, this should not be default. I would not mind adding a config field in containers.conf to enable this more minimal events output but we should not (and, as @Luap99 pointed out, cannot without a major version) do this by default.

**Comment by Luap99** [comment]
@harish2704 Our API socket is split in two parts the normal docker api endpoint and the our libpod endpoints that all start with /version/libpod/... so all the other docker compatiable endpoints can and should be changed to match docker api as closely as possible.

Look into pkg/api/handlers/compat/events.go there we use the code for both endpoints but if you look there into the logic you will find `!utils.IsLibpodRequest(r)` usage there so based on that you can change the behavior there. Now of course because the event stream cannot known when such a healthcheck state change happens you need to add this info into the event itself so that you can filter it there based on that which is why I suggest added this field or attribute to the event type.

File: libpod/healthcheck.go (hunk 1 above)
Comment (harish2704) [reply libpod/healthcheck.go]: done in https://github.com/containers/podman/pull/24005/commits/5b8d32d26bc1eebc45d82b40e9866a0459a6a500

**Comment by harish2704** [comment]
@Luap99 
I made the changes as you suggested. In short, they are 

1. Added extra flag to to Event struct to indicate whether health status changed or not
2. Marshal & parse that flag while writing to Journald log
3. Check that flag before writing/publishing an even in docker compatible API. Other APIs are not touched. 

ie, if we run `podman events` we can see the old normal behavior of too many logs 
but if we run `env DOCKER_HOST=unix://<path-to-podman.sock> docker logs` then we can not see to much logs for health check. Logs are emitted only when there is a state change

File: libpod/events/config.go [hunk 2]
```diff
@@ -59,6 +59,8 @@ type Details struct {
 	// Attributes can be used to describe specifics about the event
 	// in the case of a container event, labels for example
 	Attributes map[string]string
+	// Whether state change happened for HealthStatus or not
+	IsHealthStatusChanged bool
```
Comment (Luap99) [inline libpod/events/config.go:63]: This needs to be at least omitempty like other fields so it doesn't show up on non hc events. Maybe it would be better to add this into into the Attributes map instead, @mheon WDYT?

File: libpod/events.go [hunk 3]
```diff
@@ -28,31 +28,32 @@ func (r *Runtime) newEventer() (events.Eventer, error) {
...
+	if err := c.newContainerEventWithInspectData(status, events.EventMetadata{}, false); err != nil {
 		logrus.Errorf("Unable to write container event: %v", err)
 	}
 }
 
 // newContainerHealthCheckEvent creates a new healthcheck event with the given status
-func (c *Container) newContainerHealthCheckEvent(healthStatus string) {
-	if err := c.newContainerEventWithInspectData(events.HealthStatus, healthStatus, false); err != nil {
+func (c *Container) newContainerHealthCheckEvent(healthStatus string, isHcStatusChanged bool) {
+	if err := c.newContainerEventWithInspectData(events.HealthStatus, events.EventMetadata{HealthStatus: healthStatus, IsHealthStatusChanged: isHcStatusChanged}, false); err != nil {
```
Comment (Luap99) [inline libpod/events.go:38]: This will cause conflicts with https://github.com/containers/podman/pull/23900, not sure what is best here and which one should be merged first.

File: pkg/api/handlers/compat/events.go [hunk 4]
```diff
@@ -93,6 +93,10 @@ func GetEvents(w http.ResponseWriter, r *http.Request) {
 			if evt == nil {
 				continue
 			}
+			if evt.Status == events.HealthStatus && !evt.IsHealthStatusChanged{
```
Comment (Luap99) [inline pkg/api/handlers/compat/events.go:96]: This needs a !utils.IsLibpodRequest(r) check like the other docker compat changes below as this function is used for both the docker and libpod endpoint.

**Review by Luap99 — COMMENTED** [review]
This will need some API level tests to ensure it working, have a look at test/apiv2

File: libpod/events.go (hunk 3 above)
Comment (Honny1) [reply libpod/events.go:38]: I would merge #23900 first. I think this PR will need more care.

File: libpod/events/config.go (hunk 2 above)
Comment (harish2704) [reply libpod/events/config.go:63]: By reading the code, I understand that attribute map is fully allocated to storing container's labels 
That is why I decided not to choose that. ( Initially I thought to store this in attributes map )

Comment (harish2704) [reply libpod/events/config.go:63]: @Luap99 , @mheon : Please let me know what is the correct place to store these data. I will rebase, make changes & test my changes accordingly

**Comment by Honny1** [comment]
@harish2704 Do you still want to work on it, or can I take over?

**Comment by harish2704** [comment]
@Honny1 It will get competed quickly if any member of the podman team takeover this . So I would recommend that.

For me, i may need help/guidance from the team for identifying what to do and where to do  ( For eg: I got stuck at this https://github.com/containers/podman/pull/24005#discussion_r1780408681 )

FYI, I am actually running this particular patch in a staging ( ie, non critical ) enviroment where I host ~60 containers using Coolify , but I don't have much insights on the codebase of podman project

File: libpod/events/config.go (hunk 2 above)
Comment (Honny1) [reply libpod/events/config.go:63]: I think using attributes is a good way to do that.

File: pkg/api/handlers/compat/events.go (hunk 4 above)
Comment (Honny1) [reply pkg/api/handlers/compat/events.go:96]: I think it might be beneficial to limit writing HealthCheck events for some high-performance use. I think this reduction could be configurable in the global configuration. WDYT? @Luap99 @mheon

**Comment by Honny1** [comment]
@harish2704 I will ping them.

File: pkg/api/handlers/compat/events.go (hunk 4 above)
Comment (Luap99) [reply pkg/api/handlers/compat/events.go:96]: There is already an option to disable *all* healthcheck events, `healthcheck_events` in containers.conf. I agree the logging only when the state actually changed makes sense overall. However I don't think we can just do this as I think it would count as breaking change.

So adding a second option there could be a possibility but I like to avoid more option unless absolutely necessary. I would be fine to change such a behavior in a major version but I guess that is still far away so a new containers.conf option seems like the best solution currently.
But to fix this specific issue from @harish2704 we really don't need it, not logging to journald would be helpful but the real issue is emitting the events not like docker does. This fix here can handle it to allow the docker API to behave like docker so the direct consumers work properly.

Comment (Honny1) [reply pkg/api/handlers/compat/events.go:96]: Well, we can add a second option if there is interest from the community/customers. And change the behavior in the next major release. Is there any way to record this as a potential change for the next major release?

**Comment by harish2704** [comment]
@Luap99 , @Honny1  I have updated this PR against latest main branch. Can you please review the changes ?

**Comment by martinpitt** [comment]
This "breaks" the cockpit-podman tests in the obvious way -- instead of the expected series of positive health checks, there is now just one. I'm happy to adjust the tests if this gets a positive review from the podman authors.

**Comment by Luap99** [comment]
> This "breaks" the cockpit-podman tests in the obvious way -- instead of the expected series of positive health checks, there is now just one. I'm happy to adjust the tests if this gets a positive review from the podman authors.

Are you consuming the compat API or the libpod api event stream? Like I mentioned in https://github.com/containers/podman/pull/24005#issuecomment-2361630358 this requires the use of `!utils.IsLibpodRequest(r)` so we don't break the libpod API which would be an unacceptable break. The cockpit API test should not have to change unless we do a new major release.

**Comment by harish2704** [comment]
@martinpitt , @Luap99 : To be honest, I really forgot about the `!utils.IsLibpodRequest(r)` thing while making this PR. 
but now I did following test from end . 

1. After applying this patch, started 3 containers in which health check runs at every second.
2. then opened one terminal with`podman events` command. I was able to see  log output  at every second ( 3 output in every second for 3 containers )
3. In another terminal and  run `docker events`  command. Docker is actually communicating to podman service over podman.sock. I saw log output only when health status changes. 

I hope the above mentioned behavior indicates that this PR is not breaking libpod API. Is that correct ? 

( The compose files and other resources  I used for  testing can be found  here https://github.com/containers/podman/issues/24003#issue-2534255215 )

**Comment by martinpitt** [comment]
@Luap99 :

> > This "breaks" the cockpit-podman tests in the obvious way -- instead of the expected series of positive health checks, there is now just one. I'm happy to adjust the tests if this gets a positive review from the podman authors.
> 
> Are you consuming the compat API or the libpod api event stream?

current cockpit-podman listens to `/v1.12/libpod/events` as per https://docs.podman.io/en/stable/_static/api.html#tag/system/operation/SystemEventsLibpod . I'm not sure which of the two that is?

I suppose at some point we should bump the v1.12, but we still support  e.g. podman 3.4 in Ubuntu 22.04 LTS, or podman 4.0 in RHEL 8. TBH we haven't paid attention to bumping this, it's mostly "our tests will tell us when v1.12 stops working".

Thanks!

**Comment by harish2704** [comment]
@Luap99 : Sorry. Please ignore my review request. It was made accidentally. 
I was trying to view the change you requested, I saw a message like "[Luap99](https://github.com/Luap99)
Requested changes" but coudn't find what it is

**Comment by Luap99** [comment]
> @Luap99 :
> 
> > > This "breaks" the cockpit-podman tests in the obvious way -- instead of the expected series of positive health checks, there is now just one. I'm happy to adjust the tests if this gets a positive review from the podman authors.
> > 
> > 
> > Are you consuming the compat API or the libpod api event stream?
> 
> current cockpit-podman listens to `/v1.12/libpod/events` as per https://docs.podman.io/en/stable/_static/api.html#tag/system/operation/SystemEventsLibpod . I'm not sure which of the two that is?

Everything with `vX/libpod/` are libpod endpoints so you are using the correct one. The paths that do not start with libpod/ are consider docker compat API we try to match the docker API interface there so if there is a bug that is not right for docker we might decide to chnage that without a major version.
> 
> I suppose at some point we should bump the v1.12, but we still support e.g. podman 3.4 in Ubuntu 22.04 LTS, or podman 4.0 in RHEL 8. TBH we haven't paid attention to bumping this, it's mostly "our tests will tell us when v1.12 stops working".

Oh well that version doesn't make any real sense to us. Generally we are tracking the podman version so you should be using something like `v5.0.0`, the thing is our API versioning is a mess and we seems to not really validate the version at all.  In general we have almost no do X only if api version > Y code. There are a few cases but I doubt they would matter to cockpit, in case someone is interested to look at them just grep for `SupportedVersion(`.

The most obvious case are the network endpoints

```
$ curl --unix-socket $XDG_RUNTIME_DIR/podman/podman.sock -X GET http://d/v1.12/libpod/networks/json -H "Content Type: application/json" 
{"cause":"given version is not supported","message":"failed to parse query parameter 'version': \"1.12.0\": given version is not supported","response":400}
```

**Comment by martinpitt** [comment]
Thanks @Luap99 ! I sent https://github.com/cockpit-project/cockpit-podman/pull/2185 to bump the API version to `v3.4.0/` which is clearer indeed, even if it makes little or no practical difference.

**Comment by harish2704** [comment]
@Luap99 @martinpitt : Do I have to do anything from my end on this PR?

**Comment by Honny1** [comment]
> @Luap99 @martinpitt : Do I have to do anything from my end on this PR?

You don't have to do much, but there are a few small things to address. The CI should be passing, so let's aim for that. It looks like there are some minor linter nits that need fixing. Also, please make sure to use `!utils.IsLibpodRequest(r)` where appropriate.

**Comment by Luap99** [comment]
> @Luap99 @martinpitt : Do I have to do anything from my end on this PR?

Note @martinpitt is a maintainer of cockpit and not podman, he just comments here if he see changes that break the cockpit-podman tests which also helps us to avoid breaking them with unwanted API changes.



@harish2704 CI is not passing here, see the jobs error. The linter is complaining https://github.com/containers/podman/pull/24005/checks?check_run_id=45370701706 so that must be fixed.

I have not yet reviewed this properly but will try to do so later this week.

**Comment by Honny1** [comment]
@harish2704 Please add tests to verify the behavior for both the `Libpod` API and the `Compat` API. You can refer to the existing [Events API tests](https://github.com/containers/podman/blob/main/test/apiv2/27-containersEvents.at) as a reference.

**Comment by harish2704** [comment]
@Luap99 @Luap99  I have one question. Do i need to implement this for each log backends?

Current implementation was working when backen was jounard, but it found be broken when I switched logging to json file

**Comment by Honny1** [comment]
@harish2704 Can you please rebase on upstream main?

**Comment by harish2704** [comment]
@Honny1 I rebased . 
I also wrote a different implementation with the help of LLM . That can be seen here https://github.com/podman-container-tools/podman/compare/main...harish2704:podman:fix-docker-compatible-healthcheck-events . Take look at it if possible. Unit tests are there too

**Comment by Honny1** [comment]
I was thinking about this behavior, and I am not sure if we want this since it is a breaking change. But this behavior could be configurable, and it makes sense to me.  WDYT? Should we close this, or proceed with a config option?  @podman-container-tools/podman-maintainers

**Comment by mheon** [comment]
I am fine with this as long as it is configurable, and not the default. Log spam is a real problem, but we also cannot break default config for healthchecks.
